File: clients/dart/cldf/lib/models/sector.dart
import 'package:json_annotation/json_annotation.dart';

import 'location.dart' show Coordinates;
import 'media/media.dart';

part 'sector.g.dart';

/// A sector within a climbing location
@JsonSerializable(includeIfNull: false)
class Sector {
  /// Creates a new [Sector] instance
  Sector({
    required this.id,
    this.clid,
    required this.locationId,
    required this.name,
    this.isDefault = false,
    this.description,
    this.approach,
    this.coordinates,
    this.media,
    this.createdAt,
    this.tags,
    this.customFields,
  });

  /// Creates a [Sector] from JSON
  factory Sector.fromJson(Map<String, dynamic> json) => _$SectorFromJson(json);

  /// Unique identifier
  final int id;

  /// CrushLog ID - globally unique identifier (v1.3.0+)
  final String? clid;

  /// Reference to location
  final int locationId;

  /// Sector name
  final String name;

  /// Whether this is the default sector for the location
  final bool isDefault;

  /// Description
  final String? description;

  /// Approach information
  final String? approach;

  /// GPS coordinates
  final Coordinates? coordinates;

  /// Media associated with this sector (overview photos, approach maps, etc.)
  final Media? media;

  /// Creation timestamp
  @JsonKey(toJson: _dateTimeToJson, fromJson: _dateTimeFromJson)
  final DateTime? createdAt;

  /// Associated tags
  final List<String>? tags;

  /// Custom fields
  final Map<String, dynamic>? customFields;

  /// Converts this [Sector] to JSON
  Map<String, dynamic> toJson() => _$SectorToJson(this);
}

// Helper functions for DateTime serialization
String? _dateTimeToJson(DateTime? dateTime) =>
    dateTime?.toUtc().toIso8601String();
DateTime? _dateTimeFromJson(String? json) =>
    json == null ? null : DateTime.parse(json);

File: clients/java/cldf-java/src/main/java/app/crushlog/cldf/models/Sector.java
package app.crushlog.cldf.models;

import java.time.OffsetDateTime;

import app.crushlog.cldf.models.media.Media;
import app.crushlog.cldf.utils.FlexibleDateTimeDeserializer;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
/** Represents a Sector in the CLDF archive. */
public class Sector {

  @JsonProperty(required = true)
  private Integer id;

  /** CrushLog ID - globally unique identifier (v1.3.0+) */
  private String clid;

  @JsonProperty(required = true)
  private Integer locationId;

  @JsonProperty(required = true)
  private String name;

  @JsonProperty("isDefault")
  @Builder.Default
  private boolean isDefault = false;

  private String description;

  private String approach;

  private Coordinates coordinates;

  /** Media associated with this sector (overview photos, approach maps, etc.) */
  private Media media;

  @JsonDeserialize(using = FlexibleDateTimeDeserializer.class)
  private OffsetDateTime createdAt;

  @Data
  @Builder
  @NoArgsConstructor
  @AllArgsConstructor
  public static class Coordinates {
    @JsonProperty(required = true)
    private Double latitude;

    @JsonProperty(required = true)
    private Double longitude;
  }
}
